backend/audit/admin.py

```diff
@@ -24,9 +27,73 @@
 )
 from django.contrib.admin import SimpleListFilter
 from django.utils.translation import gettext_lazy as _
+from django.utils.timezone import now
+
+from dissemination.remove_singleauditreport_pdf import remove_singleauditreport_pdf
+from dissemination.remove_workbook_artifacts import remove_workbook_artifacts
 
 logger = logging.getLogger(__name__)
 
+# As per ADR #0041, the retention period for flagged reports is 6 months. That is 180 days.
+FLAGGED_REPORT_RETENTION_DAYS = 180  # 6 months
+
+
+@admin.action(description="Delete reports flagged for removal for over 6 months")
+def delete_flagged_records(modeladmin, request, queryset):
+    """
+    Admin action to delete records flagged for removal older than the specified months.
+    """
+    cutoff_date = now() - datetime.timedelta(days=FLAGGED_REPORT_RETENTION_DAYS)
+
+    # Filter records for deletion
+    records_to_delete = queryset.filter(
+        submission_status=STATUS.FLAGGED_FOR_REMOVAL,
+    ).exclude(
+        transition_name__isnull=True,
+        transition_date__isnull=True,
+    )
+
+    filtered_records = []
+    for sac in records_to_delete:
+        try:
+            # Check if the last transition state is FLAGGED_FOR_REMOVAL
+            if sac.transition_name[-1] == STATUS.FLAGGED_FOR_REMOVAL:
+                # Get the corresponding transition_date
+                transition_datetime = sac.transition_date[-1]
+
+            if transition_datetime <= cutoff_date:
+                filtered_records.append(sac)
+        except (ValueError, IndexError, TypeError) as e:
+            logger.error(
+                f"Error filtering flagged records for deletion: {str(e)}",
+                exc_info=True,
+            )
+            continue
+
+    count = 0
+    report_ids = []
+    for sac in filtered_records:
+        auditee_uei = sac.general_information.get("auditee_uei")
+        if auditee_uei:
+            UeiValidationWaiver.objects.filter(uei=auditee_uei).delete()
+        SacValidationWaiver.objects.filter(report_id=sac).delete()
+        remove_singleauditreport_pdf(sac)
+        remove_workbook_artifacts(sac)
+        Access.objects.filter(sac=sac).delete()
+        SubmissionEvent.objects.filter(sac=sac).delete()
+        ExcelFile.objects.filter(sac=sac).delete()
+        SingleAuditReportFile.objects.filter(sac=sac).delete()
+        report_ids.append(sac.report_id)
+        sac.delete()
```

Comment:
I am still getting caught up to speed on Django ORM...

It looks like we iterate over all the SAC records, then one by one delete entries with a foreign key relation, followed by deleting the single sac.

Does Django handle all those delete as a single transaction? If no, I am a bit concerned about failures in the middle of this process. For example, what if the SubmissionEvent fails to delete, won't that leave us in an inconsistent state?

If the database has cascading deletes on all the foreign key constraints, we could get away with just deleting all the sacs at once rather than in loop.

Reply:
Placing this process under a transaction is generally a good idea to avoid unexpected issues. The current implementation, even without a global transaction, will not leave the database in an inconsistent state. This is because the sac record, which acts as the parent for all the other records in the list, is only removed at the end of the process. If an unexpected issue occurs midway through the process, the sac record will remain intact. When the action is re-run, it will clean up any remaining child records before finally removing the sac. As long as the parent sac record is not removed prematurely, the child records can be handled in subsequent runs of the deletion action.
That said, I still add a global transaction as an extra layer of protection.